arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -48,17 +48,40 @@ SECTION_FUNC(TEXT, z_arm_cpu_idle_init)
 #endif
 	bx	lr
 
-SECTION_FUNC(TEXT, arch_cpu_idle)
-#ifdef CONFIG_TRACING
+.macro _prologue
+#if defined(CONFIG_TRACING) || defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
 	push	{r0, lr}
+#if defined(CONFIG_TRACING)
 	bl	sys_trace_idle
+#endif
+#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
+	bl	z_arm_on_enter_cpu_idle
+	/* Preserve the return value so that it can be used later,
+	 * see the _sleep_if_allowed macro.
+	 */
+	mov	r2, r0
+#endif
```

Comment:
The amount of duplicate code here is very minimal. IMHO, it is better to not make this a macro since doing so makes code harder to follow.

Reply:
I got rid of the `_prologue` macro.